text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+
+
+# Proposed techniques
+[technique]: #proposed-technique
+
+The proposed approach is heavily inspired by SmartBFT but is designed to be consensus agnostic, minimize implementation and operational complexity.
+
+### Block replication
+
+The proposed technique for block replication is essentially the approach of SmartBFT where a re-shuffling takes place not only upon suspicion of the block source but also in a timely fashion in order to be resilient to overloaded or deliberately slow orderers.   
+In order to save network bandwidth and processing power when the network is idle, the automatic time-based re-shuffling will only take place if new blocks are received.
+Moreover, there will not be a **need** to specify in the peer's local configuration (or an orderer's local configuration) which of the implementations to use. 
+Notice that a block validation policy that mandates signatures from multiple ordering nodes, is **never** satisfied by **only** principal sets of size one. 
+Or, equivalently, a block validation policy that is satisfied by **only** principal sets of size one never requires signatures from multiple ordering nodes. 
+Evidently, policies of the former type are **always** BFT policies, while policies of the latter type might be either a crash fault policy or a BFT policy that utilizes a threshold signature scheme.
+
+Therefore, Fabric nodes will pick the BFT implementation if either the block validation policy implies the need, or if an override is activated in the node's local configuration.  
+
+
+### Block signature verification
+
+The proposed technique for block signature verification should be generic and consensus agnostic, in order to facilitate incorporation of multiple consensus types without the need for Fabric nodes to understand which consensus implementation is employed.
+
+Apart from the already existing block validation policy:
+
+```
+# BlockValidation specifies what signatures must be included in the block
+# from the orderer for the peer to validate it.
+BlockValidation:
+    Type: ImplicitMeta
+    Rule: "ANY Writers"
+```
+
+The `Orderer` section in the channel config will contain a mapping from numerical identifiers to tuples of serialized identities:
+
+```
+Orderer: &OrdererDefaults
+    ...
+    ConsenterMapping:
+    - Id: 1
+      MSPID: OrdererOrg1
+      Identity: <base64 encoding of raw identity>
+    - Id: 2
+      MSPID: OrdererOrg2
+      Identity: <base64 encoding of raw identity>
+      ...
+```
+
+Upon a configuration update, it will be up to the (Fabric side of the) consensus implementation to validate that:
+
+- The block validation policy matches the required threshold (e.g `2f+1`) of the consensus protocol.
+- The `ConsenterMapping` is not in conflict with the opaque consensus metadata.
```

Comment:
What exactly is the "opaque consensus metadata"?

Reply:
It's an encoding of the consensus-specific configuration in the orderer section of the config block. It contains [these items](https://github.com/hyperledger/fabric-protos/blob/main/orderer/etcdraft/configuration.proto#L14-L17) in Raft and similarly [these items](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/orderer/smartbft/configuration.proto#L14-L17) in SmartBFT.